libfwupdplugin/generate-struct.py

```diff
@@ -375,25 +416,49 @@ def __str__(self) -> str:
         return tmp
 
 
-# convert a CamelCase name into snake_case
-def _camel_to_snake(name: str) -> str:
+class Generator:
+    def __init__(self, basename) -> None:
+        self.basename: str = basename
 
-    name_snake: str = ""
-    for char in name:
-        if char.islower() or char.isnumeric():
-            name_snake += char
-            continue
-        if name_snake:
-            name_snake += "_"
-        name_snake += char.lower()
-    return name_snake
+    def _process_enums(self, name: str, items: List[EnumItem]) -> Tuple[str, str]:
 
+        name_snake = _camel_to_snake(name)
 
-class Generator:
-    def __init__(self, basename):
-        self.basename: str = basename
+        # header
+        str_h = ""
+        str_h += "\ntypedef enum {\n"
+        for item in items:
+            if item.default:
+                str_h += f"    {item.c_define(name)} = {item.default},\n"
+            else:
+                str_h += f"    {item.c_define(name)},\n"
+        str_h += f"}} Fu{name};\n"
+        str_h += f"const gchar *fu_enum_{name_snake}_to_string(Fu{name} val);\n"
+        str_h += f"Fu{name} fu_enum_{name_snake}_from_string(const gchar *val);\n"
 
-    def _process_items(self, name: str, items: List[Item]) -> Tuple[str, str]:
+        # code
+        str_c = ""
+        str_c += "\nconst gchar *\n"
+        str_c += f"fu_enum_{name_snake}_to_string(Fu{name} val)\n"
+        str_c += "{\n"
+        for item in items:
+            str_c += f"    if (val == {item.c_define(name)})\n"
+            str_c += f'        return "{item.value}";\n'
+        str_c += "    return NULL;\n"
+        str_c += "}\n"
+        str_c += f"\nFu{name}\n"
+        str_c += f"fu_enum_{name_snake}_from_string(const gchar *val)\n"
+        str_c += "{\n"
+        for item in items:
+            str_c += f'    if (g_strcmp0(val, "{item.value}") == 0)\n'
+            str_c += f"        return {item.c_define(name)};\n"
+        str_c += f"    return {items[0].c_define(name)};\n"
+        str_c += "}\n"
```

Comment:
having seen how well jinja worked for templating the man pages recently, do you think it's worth trying here too?  It might be more readable at least as a result.

Reply:
Yes, good idea. I'll try doing this in the next PR.

Reply:
Looks like a win to me:

	 fu-rustgen-enum.c.in   |   19 +++
	 fu-rustgen-enum.h.in   |   11 ++
	 fu-rustgen-struct.c.in |  212 ++++++++++++++++++++++++++++++++++++++++++++
	 fu-rustgen-struct.h.in |   45 +++++++++
	 rustgen.py             |  390 ++++++---------------------------------------------------------------------------
	 5 files changed, 317 insertions(+), 360 deletions(-)

-- i've added it to this PR.